### ISSUE TITLE
[cssom] [css-conditional] Extend CSSStyleDeclaration to allow for nested at-rules

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#5172]
I strongly suspect this is a duplicate, but I can't find anything, so I'm gonna go ahead and post, just in case it hasn't been discussed before.

At-rules nested inside CSS rules have been proposed many times (@apply, @nest etc). In addition, nesting @supports and media queries inside CSS style rules would be incredibly useful and would allow things like

```css
#foo {
	width: 10em;
	@supports(width: min(1em, 1px)) {
		width: max(0em, 10em - var(--scrolltop));
	}
}
```

which would allow using fallbacks and MQ overrides more succinctly and keep everything in the same place.

The more CSS variables are adopted, the more `@supports` will need to be utilized for overriding fallbacks, since the cascade can't be used for fallbacks anymore. This leads to a lot of repetitive code, and a lot of disconnected code that one needs to hunt down to understand how the stylesheet works. Something like this would at least allow for keeping selectors DRY. 

Parsing-wise, @tabatkins has often proposed at-rules that are nested on the same level as CSS declarations, so I imagine it's fine syntax-wise and something that needs to happen anyway on the syntax level, even with restrictions on the type of at-rules that can be nested.

**Comment by tabatkins** [comment]
Yes, it's fine syntax-wise, it just needs (a) definition in the grammar, and more importantly (b) exposure in CSSOM.

**Comment by LeaVerou** [comment]
@tabatkins Does it's fine mean it's already allowed in css-syntax or that it's no problem to add?

**Comment by tabatkins** [comment]
Already allowed in Syntax. Every block allows at-rules inside of it, they just have to decide whether they also allow declarations or qualified rules (since they clash syntactically).

**Comment by LeaVerou** [comment]
I suppose I can remove css-syntax from the issue then.

**Comment by jonathantneal** [comment]
Thanks for opening this thread, @LeaVerou. I have not found another thread. I think this issue can address [Issue #1 from the specification](https://drafts.csswg.org/css-nesting/#cssom):

> **ISSUE 1**
> 1. Add an interface for the `@nest` rule.
> 2. Tie into the general work needed to let rules be nested into style rules.

As @tabatkins pointed out, at-rules are already allowed syntactically, as part of the algorithm to [consume a list of **declarations**](https://drafts.csswg.org/css-syntax/#consume-list-of-declarations). Does this algorithm mean that **at-rules** (and soon **nesting-rules**) should be considered kinds of **declarations**? Knowing how to refer to these things will help me contribute ideas for the interface.

Are there existing ideas for how or where these nesting-rules will fit into the existing [CSS Object Model](https://drafts.csswg.org/cssom/#css-rules)?

Should the CSSOM interface for nesting rules be addressed as part of the resolution for #5649 to “_resolve sep interfaces for declarations that declare different things_”? (Perhaps this goes back to whether these nesting rules are declarations.)

**Comment by argyleink** [comment]
this can close because the feature was [added to the spec here](https://drafts.csswg.org/css-nesting-1/#conditionals) yeah?
